File: day1-bonus.py
import utils

def __main__():
    lines = utils.read_input()
    nums = [int(line) for line in lines]
    target = 2020

    print(three_sum(nums, target))

def three_sum(nums: list[int], target: int) -> int:
    seen = {num: i for i, num in enumerate(nums)}

    for i, num in enumerate(nums):
        for j in range(i + 1, len(nums)):
            num2 = nums[j]
            num3 = target - num - num2
            
            num3Idx = seen.get(num3)
            if num3Idx is not None and num3Idx != i and num3Idx != j:
                return num * num2 * num3

__main__()

File: src/day1_bonus.rs
mod utils;

fn main() {
    let lines = utils::read_input();
    let nums: Vec<i32> = lines
        .iter()
        .map(|line| line.parse().expect("Failed to parse number"))
        .collect();
    let target = 2020;

    println!("{}", three_sum(&nums, target));
}

fn three_sum(nums: &[i32], target: i32) -> i32 {
    let seen: std::collections::HashMap<i32, usize> = nums
        .iter()
        .enumerate()
        .map(|(i, &num)| (num, i))
        .collect();

    for (i, &num) in nums.iter().enumerate() {
        for j in (i + 1)..nums.len() {
            let num2 = nums[j];
            let num3 = target - num - num2;

            if let Some(&k) = seen.get(&num3) {
                if k != i && k != j {
                    return num * num2 * num3;
                }
            }
        }
    }

    panic!("No solution found")
}
